### ISSUE TITLE
Faking the timezone

### ISSUE THREAD

**Issue by SystemParadox** [sinonjs/lolex#48]
How much work would it be to support faking the timezone? Is this even possible?

Over at node-schedule there are repeated issues due to daylight savings time changes (see https://github.com/node-schedule/node-schedule/issues/267). I was just about to write some tests when I realised that the tests will only work (or rather... only fail!) if the person running the tests also has their timezone set to the same localtime as me. If the timezone is different, the bug will occur at different times. If they are using UTC, the bug will not occur at all.

So I would like to be able use lolex and tell it to pretend we are in a particular timezone, regardless of what the system timezone actually is. My suspicion is that this may not even be possible.

Any thoughts?

Thanks.

**Comment by mantoni** [comment]
You can test code independent of the system time and factor in timezones. This is just a matter of how you construct the date object. Consisted this:

```js
new Date('2016-04-01Z11:58:00+01:00').toString()
  == 'Fri Apr 01 2016 12:58:00 GMT+0200 (CEST)'
new Date('2016-04-01Z11:58:00+01:00').toISOString()
  == '2016-04-01T10:58:00.000Z'
```

So you can see that JavaScript deals with the timezone offset properly. 

However, Lolex only deals with milliseconds. It has no internal concept of timezones and I’m not sure whether this is within Lolex’ scope. You can initialize the Lolex clock instance with a specific timestamp like this:

```js
var now = ew Date('2016-04-01Z11:58:00+01:00').getTime();
var clock = lolex.createClock(now);
```

This will fake the current Date to behave as if it was in the specified timezone offset. You can then use the `clock` API as usual to advance the time with `tick` etc.

Does that help solve your problem?

**Comment by SystemParadox** [comment]
For me, with my timezone set to `Europe/London`, I get the following:

```
new Date('2016-04-01Z11:58:00+01:00').toString()
  == 'Fri Apr 01 2016 11:58:00 GMT+0100 (BST)'
```

So this is different to your output, because we have our timezones set differently. This is what I am trying to resolve.

It probably is out of scope for lolex, although any package which did allow this would also have to implement lolex' functionality, unless there was some way for them to cooperate despite both needing to modify `Date`.

**Comment by mantoni** [comment]
Ah, I understand. You can set the `TZ` environment variable to set a specific timezone. This works in node and phantomjs like this:

```bash
$ export TZ='Europe/London‘
$ echo "console.log(new Date('2016-04-01').toString())" | node
Fri Apr 01 2016 01:00:00 GMT+0100 (BST)
$ echo "console.log(new Date('2016-04-01').toString())" | phantomjs
phantomjs> Fri Apr 01 2016 01:00:00 GMT+0100 (BST)
```

I can see how this is inconvenient if you want to test multiple timezone settings in a single test file.
Lolex already mirrors native date properties [like toString](https://github.com/sinonjs/lolex/blob/master/src/lolex-src.js#L129) so this might be a good place to implement fake timezones. I think it’s entirely possible, but it’s not super easy to get right.

**Comment by mantoni** [comment]
This might help with an implementation: http://momentjs.com/timezone/

**Comment by rogierschouten** [comment]
@mantoni 's comment about the TZ environment variable is correct, however I found that there are differences in Date behaviour on systems with different time zones even if TZ is set the same.  
Also, the Date class behaves differently in Node.JS, IE, FF, Chrome etc.
While lolex could fake that by replacing the entire Date class, it is very hard to know how to correctly simulate the platform. 

This is why I created the timezonecomplete package.  I tried moment.js at first but at the time it had some bugs around DST changes.
